cassiopeia/core/league.py

```diff
@@ -156,15 +162,24 @@ def platform(self) -> Platform:
 
     @property
     def league_id(self) -> str:
-        return self._data[LeagueEntryData].leagueId
+        try:
+            return self._data[LeagueEntryData].leagueId
+        except AttributeError:
+            return self.league.id
 
     @lazy_property
     def queue(self) -> Queue:
-        return Queue(self._data[LeagueEntryData].queue)
+        try:
+            return Queue(self._data[LeagueEntryData].queue)
+        except AttributeError:
+            return self.league.queue
 
     @property
     def name(self) -> str:
-        return self._data[LeagueEntryData].name
+        try:
+            return self._data[LeagueEntryData].name
```

Comment:
I'm almost certain that one of the endpoints that returns a LeagueEntryData object contains a `name` field. In this case, if the LeagueEntry is unloaded and the `.name` field is called, the data will not be pulled from the Riot API. Instead, `self.league.name` will be requested, which is the incorrect call.

I don't have my head completely wrapped around the most recent changes to the League endpoints, but I'm pretty sure this try/except behavior can cause Cass to make a call to the wrong endpoint.

Reply:
The way I understand it, is that `League` encapsulates `LeagueListDTO`, and `LeagueEntry` encapsulates both `LeagueItemDTO` (which is contained only in `LeagueListDTO`) and `LeagueEntryDTO`. The only data structure that gives the name of the league is `LeagueListDTO`. Here are the returns of all v4/league endpoints, with type annotated in angle brackets:

`/lol/league/v4/leagues/{leagueId}`:

```
<LeagueListDTO>{
    "tier": "SILVER",
    "leagueId": "58d9b5e0-2031-11e9-b995-c81f66cf2333",
    "queue": "RANKED_SOLO_5x5",
    "name": "Wukong's Heralds",
    "entries": [
       <LeagueItemDTO> {
            "summonerId": "mQZM5iaazyV5a-BtotogdrrXNQuzel-lcGE3iU1zFcw3sZA",
            "summonerName": "ChanceCSGO",
            "leaguePoints": 23,
            "rank": "III",
            "wins": 47,
            "losses": 42,
            "veteran": false,
            "inactive": false,
            "freshBlood": false,
            "hotStreak": false
        },
        ...
    ]
}
```

`challengerleagues/masterleagues/grandmasterleagues`
```
{
    "tier": "CHALLENGER",
    "leagueId": "974b70e3-28eb-3b60-9e9f-82a8efa19f10",
    "queue": "RANKED_SOLO_5x5",
    "name": "Syndra's Masterminds",
    "entries": [
        <LeagueItemDTO>{
            "summonerId": "7W_HUtkYtWSH23bEx03P2uL5c1NGkRpRssn8-f3ma_e3gvU",
            "summonerName": "Humble Diligent",
            "leaguePoints": 972,
            "rank": "I",
            "wins": 244,
            "losses": 184,
            "veteran": false,
            "inactive": false,
            "freshBlood": false,
            "hotStreak": false
        },
        ...
    ]
}
```


`/lol/league/v4/entries/{queue}/{tier}/{division}`
```
[
    <LeagueEntryDTO>{
        "leagueId": "b30e9d61-ab9f-11e9-809c-c81f66cf2333",
        "queueType": "RANKED_SOLO_5x5",
        "tier": "DIAMOND",
        "rank": "I",
        "summonerId": "5ercERfLOuNhuENu6Egh7CXKKeTHxpxCpRUiHRu_51de_4nE",
        "summonerName": "xTc4",
        "leaguePoints": 100,
        "wins": 224,
        "losses": 209,
        "veteran": false,
        "inactive": false,
        "freshBlood": false,
        "hotStreak": false,
        "miniSeries": {
            "target": 3,
            "wins": 0,
            "losses": 1,
            "progress": "LNNNN"
        }
    },
    ...
]
```

`/lol/league/v4/entries/by-summoner/{encryptedSummonerId}`
```
[
    <LeagueEntryDTO>{
        "leagueId": "19dced70-ab60-11e9-9e48-c81f66cf135e",
        "queueType": "RANKED_TFT",
        "tier": "BRONZE",
        "rank": "I",
        "summonerId": "bO_4jcBLPekTQiB0vlUS1qXF-zB-4_7CTRMIHayR1SEmit8",
        "summonerName": "Kalturi",
        "leaguePoints": 20,
        "wins": 2,
        "losses": 27,
        "veteran": false,
        "inactive": false,
        "freshBlood": false,
        "hotStreak": false
    },
    ...
]
```

I am sorry if I am wasting your time with my misunderstanding of what corresponds to what, it very well could be that I am writing this code and comments with completely incorrect assumptions, but it seems that the only place to actually get the league name is `/lol/league/v4/leagues/{leagueId}` or the challengerleagues/masterleagues/grandmasterleagues endpoints, so I'm not sure how it could ever be incorrect to lookup the league name from the class that encapsulates the response from `/lol/league/v4/leagues/{leagueId}` if the league name is not loaded through code in `LeagueEntry` construction

Reply:
I'm finally getting to this. Thanks for posting all the details.

So it looks like `LeagueEntry`-like objects (which I'm defining as objects that contain information about the league of a specific summoner) never have the `leagueName` field, which you pointed out. Instead, the league name is only available when requesting a league via it's ID, and (maybe coincidentally) when requesting challenger/grandmaster/master.

Because of that, my gut reaction is that the League should be treated like the Summoner for these LeagueEntry objects. I.e. each LeagueEntry corresponds to a summoner, so it has a `.summoner` attribute that returns a Summoner object; and similarly, each LeagueEntry corresponds to a league, so it has a `.league` attribute that returns a League object. There is a difference: the LeagueEntry and Summoner objects are (for a give queue) one-to-one, whereas the LeagueEntry and League objects are many-to-one. I think that's fine though. It's a little weird that you'd be able to access the "parent" object of a LeagueEntry by hitting the `.league` attribute, however. I'm not sure I like that.

I'm a bit up in the air right now about this. It seems like there are two options:

Option 1: Each `LeagueEntry` has a `.league_name` and `.league_id` attribute.

Option 2: Each `LeagueEntry` has a `.league` attribute which in turn has a `.name` and `.id` attribute.

Option 2 seems more "correct" because the `.name` and `.id` are on the correct object (the `League` rather than the `LeagueEntry`). On the other hand, it's a bit strange to be accessing the "parent" object from the child object since, technically, every `LeagueEntry` belongs to a given `League`.

Reply:
Once we get this fleshed out, everything else looks good.

Reply:
I'm not sure what's so bad about being able to access a parent data object from the class, any languages standard library TreeNode implementation has a get parent function, any sort of XML/HTML parsing functionality will also easily let you get the parent from any child node, etc. , it seems like a perfectly natural way to organize and retrieve data that has a strictly hierarchical format like Leagues->LeagueEntries

As for the difference between the two options, it's essentially a question of cass should closer mimic riots data definitions or if it should make it easy and intuitive to access fields where users think they should be, which is why I think it's better from a design point of view to have both a reference to the parent as well as alias properties on the LeagueEntry itself for accessing common things about the league (which is how it was in the previous version of the API anyway if I'm not mistaken)

Reply:
We're going to go with this:

```
class LeagueEntry:
    @property .league

class League:
    @property .id
    @property .name
```

so the `LeagueEntry` object won't have a `.league_id` or `.league_name`.

Note that it's crucial that `LeagueEntry.league.id` does NOT make a call to the Riot API.

If you want to make those changes I'll look through things again. Not sure how difficult it is to do that. Ideally we'll be able to keep things very simple. But this should take care of the design decisions, and now the bottleneck will be implementation, so I'm close to happy with merging this.